### PR TITLE
Online DDL: Fail a --in-order-completion migration, if a _prior_ migration within the same context is 'failed' or 'cancelled'

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/16070

Either at scheduling time, or when `running`, for a given `--in-order-completion` migration, if we find another migration in the same `migration-context` that is `failed` or `cancelled`, _and_ that other migration _precedes_ our given migration chronologically (has a lower `id` value in `schema_migrations` table), then we fail our given migration.

Basically the idea is: if 5 migrations are meant to run in-order, and 3rd migration happens to fail, then 4th and 5th migrations should fail as well and should not be allowed to complete. If they were allowed to complete, they'd run out of order.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/16070

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go [hunk 1]
```diff
@@ -1183,6 +1191,36 @@ func testScheduler(t *testing.T) {
...
+			fmt.Sprintf("drop view  if exists t1_%s", u),
+			fmt.Sprintf("drop table if exists t2_%s", u),
+			fmt.Sprintf("drop view  if exists t3_%s", u),
+		}
+		sql := strings.Join(sqls, ";")
+		var vuuids []string
+		t.Run("drop multiple tables and views, in-order-completion", func(t *testing.T) {
+			uuidList := testOnlineDDLStatement(t, createParams(sql, ddlStrategy+" --in-order-completion", "vtctl", "", "", true)) // skip wait
+			vuuids = strings.Split(uuidList, "\n")
+			assert.Equal(t, 4, len(vuuids))
```
Comment (mattlord) [inline go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go:1209]: If you didn't already know, you can use `assert.Len` too/instead.

**Review by mattlord — APPROVED** [review]

File: go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go (hunk 1 above)
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go:1209]: I did not know!

Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/scheduler/onlineddl_scheduler_test.go]: done

File: go/vt/vttablet/onlineddl/executor.go [hunk 2]
```diff
@@ -3384,6 +3426,58 @@ func (e *Executor) executeMigration(ctx context.Context, onlineDDL *schema.Onlin
...
+		// which have failed.
+		if onlineDDL.StrategySetting().IsInOrderCompletion() {
+			wasFailed, err := e.validateInOrderMigration(ctx, onlineDDL)
+			if err != nil {
+				return nil, err
+			}
+			if wasFailed {
+				continue
+			}
+		}
```
Comment (shlomi-noach) [inline go/vt/vttablet/onlineddl/executor.go:3472]: Most of this function is just an extract/refactor out of `runNextMigration()` and into an independent function of its own. But these lines above are then also added to fail a migration before even running it, if the in-order terms are not met.

**Review by dbussink — APPROVED** [review]

**Pull request merged**